controllers/keystoneapplicationcredential_controller.go

```diff
@@ -0,0 +1,517 @@
+	}
+
+	logger.Info("Created AC in Keystone",
+		"ACID", created.ID,
+		"userID", userID,
+		"expiresAt", created.ExpiresAt,
+	)
+	return created.ID, created.Secret, created.ExpiresAt, nil
+}
+
+func (r *ApplicationCredentialReconciler) storeACSecret(
+	ctx context.Context,
+	helperObj *helper.Helper,
+	ac *keystonev1.KeystoneApplicationCredential,
+	secretName, newID, newSecret string,
+) error {
+
+	data := map[string]string{
+		"AC_ID":     newID,
+		"AC_SECRET": newSecret,
+	}
+
+	// Add application-credentials label
+	labels := map[string]string{
+		"application-credentials": "true",
+	}
+
+	tmpl := []util.Template{{
+		Name:       secretName,
+		Namespace:  ac.Namespace,
+		Type:       util.TemplateTypeNone,
+		CustomData: data,
+		Labels:     labels,
+	}}
+	if err := oko_secret.EnsureSecrets(ctx, helperObj, ac, tmpl, nil); err != nil {
+		return err
+	}
+
+	// Ensure protection finalizer is present on the Secret
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{Namespace: ac.Namespace, Name: secretName}
+	if err := helperObj.GetClient().Get(ctx, key, secret); err != nil {
+		return err
+	}
+	if !controllerutil.ContainsFinalizer(secret, acSecretFinalizer) {
+		base := secret.DeepCopy()
+		controllerutil.AddFinalizer(secret, acSecretFinalizer)
+		if err := helperObj.GetClient().Patch(ctx, secret, client.MergeFrom(base)); err != nil {
+			if k8s_errors.IsConflict(err) {
+				return fmt.Errorf("%w: requeue: patch conflict on %s", err, secretName)
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getUserIDFromToken extracts the user ID from the authenticated token
+func (r *ApplicationCredentialReconciler) getUserIDFromToken(_ logr.Logger, identClient *gophercloud.ServiceClient, username string) (string, error) {
+
+	// Get the authenticated user from the token
+	user, err := tokens.Get(identClient, identClient.TokenID).ExtractUser()
+	if err != nil {
+		return "", fmt.Errorf("failed to extract user %s from token: %w", username, err)
+	}
+
+	if user.ID == "" {
+		return "", fmt.Errorf("%w: user ID for user %s", util.ErrNotFound, username)
+	}
+
+	return user.ID, nil
+}
+
+// computeNextRequeue schedules the next check to happen at "expiry - grace"
+func (r *ApplicationCredentialReconciler) computeNextRequeue(ac *keystonev1.KeystoneApplicationCredential) time.Duration {
+	// default requeue is 24h as minimal grace period is 1 day
+	defaultRequeue := 24 * time.Hour
+	if ac.Status.ExpiresAt == nil || ac.Status.ExpiresAt.IsZero() {
+		return defaultRequeue
+	}
+
+	expiry := ac.Status.ExpiresAt.Time
+
+	grace := time.Duration(ac.Spec.GracePeriodDays) * 24 * time.Hour
+	rotateAt := expiry.Add(-grace)
+
+	if rotateAt.Before(time.Now()) {
+		// already within the grace window, so trigger rotation immediately
+		return 0
+	}
+
+	// otherwise check again in 24 hours
+	return defaultRequeue
+}
+
+// needsRotation returns (shouldRotate, logMessage)
```

Comment:
maybe here we can explain better the returned parameters.
In addition I'm not sure you need both `bool` and `string`.
The string seems something that feeds the logger, while the boolean is what you need to process for real as a returned value, right? In any case, I'm not asking to change this function, I was just trying to see it in the picture of the whole flow.